### PR TITLE
cgroup-support: allow to hide cgroupv2 warning via ENV

### PR DESCRIPTION
**Pull request by slyon**
Allow to hide the cgroup v2 warning by passing any value to `SNAPD_HIDE_CGROUPV2_WARNING`.
This warning to stderr makes systemd autopkgtests (and probably many others) fail and it would be handy if we could somehow disable it.

### CODE REVIEW

**Review by mardy — APPROVED** [review]
Code-wise, this is good. I'm wondering, though, because it's unusual to add an environment variable just to suppress a warning. I'd rather change the tests to filter out this line.

But let's see what others think. I'm setting this to "approved" not to block this, in case it has been agreed to be the best solution with other team members.

**Comment by slyon** [comment]
I tried filtering it out in the tests before, but that didn't work as it is buried many levels deep into the test-setup. In the case of the systemd autopkgtest the primary test is executed in a qemu VM, inside that VM LXD is being installed (via snap) and used to setup an image to run some additional (nested) autopkgtests inside containers.

snapd prints this warning to stderr and the inner (nested) autopkgtest process errors out as it sees some data on stderr. We do not have control over that nested process as it is started via autopkgtest itself. IMO it could be filtered by changing autopkgtest (in 2-3 places) to filter out this message. But it does not seem appropriate to me to modify the autopkgtest application for a snapd specific quirk.

This warning is supposed to go away soon via #10575 anyway and would be only a temporary patch to be able to run systemd autopkgtests and allow it to migrate from -proposed.

File: cmd/libsnap-confine-private/cgroup-support.c
Comment (anonymouse64) [inline cmd/libsnap-confine-private/cgroup-support.c:83]: I would prefer to see this use `getenv_bool` instead

**Review by anonymouse64 — COMMENTED** [review]
One change, otherwise the idea is fine to me temporarily. Could you also open a PR based on top of this branch which reverts this commit so that after cgroupsv2 support has landed we can merge that to remove this support? Less code in snap-confine is better IMHO

File: cmd/libsnap-confine-private/cgroup-support.c
Comment (slyon) [reply cmd/libsnap-confine-private/cgroup-support.c]: ACK, that makes a lot of sense!

**Comment by slyon** [comment]
Thank you for the review! I fixed up the `getenv_bool()` usage and agree that less code is better in snap-confine. I prepared. #10615 to revert this temporary change once it's not needed anymore.

File: cmd/libsnap-confine-private/cgroup-support.c
Comment (anonymouse64) [inline cmd/libsnap-confine-private/cgroup-support.c:100]: hide_warning is no longer a pointer to dynamically allocated memory?

**Comment by anonymouse64** [comment]
@alexmurray I think this PR is extremely simple and thus probably doesn't need a security review, but since it is touching snap-confine, I would like to get a security review for it if you (or someone else on the security) has a chance to quickly review this. Thanks

File: cmd/libsnap-confine-private/cgroup-support.c
Comment (slyon) [reply cmd/libsnap-confine-private/cgroup-support.c]: Ugh... :man_facepalming: of course not. Fixed.

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for fixing this up

**Comment by slyon** [comment]
Ouch, we have a build failure here, as the `getenv_bool()` function is not exported in the `utils.h´ header but `static` to the `utils.c` module...
Do we want to make that function available to outside code, or rather move the decision making about this environment variable into utils.c?

**Comment by anonymouse64** [comment]
> Do we want to make that function available to outside code

This makes sense to me

File: cmd/libsnap-confine-private/cgroup-support.c
Comment (alexmurray) [inline cmd/libsnap-confine-private/cgroup-support.c:83]: Would it be simpler to not add an additional `hide_warning` variable and instead just initialise `did_warn` with the result of `getenv_bool()`?

ie.

```c
static bool did_warn = getenv_bool("SNAPD_HIDE_CGROUPV2_WARNING", false);
```

**Review by alexmurray — APPROVED** [review]
LGTM!

File: cmd/libsnap-confine-private/cgroup-support.c
Comment (slyon) [reply cmd/libsnap-confine-private/cgroup-support.c:83]: Thanks for the security review!
That would be simpler, but we can only initialize the `static bool did_warn` variable with a constant value, which the `getenv_bool()` function isn't. (It would throw an `initializer element is not constant` error otherwise.)

**Comment by slyon** [comment]
@anonymouse64 OK, I think we should be good now. I handled all of your and Alex's comments and verified that the code builds locally. Sorry for this taking much longer than expected..

**Comment by anonymouse64** [comment]
The errors here are unrelated, @mvo5 / @pedronis can one of you force merge this?

**Comment by anonymouse64** [comment]
gentle ping to @mvo5 or @pedronis to land this

**Pull request merged**
